polar-core/src/validations.rs

```diff
@@ -75,11 +23,25 @@ impl<'kb> SingletonVisitor<'kb> {
     }
 
     fn warnings(self) -> Vec<Diagnostic> {
-        let mut singletons = self.singletons.into_values().flatten().collect::<Vec<_>>();
-        singletons.sort_by_key(Term::offset);
+        let mut singletons = self
+            .singletons
+            .into_iter()
+            .flat_map(|(sym, term)| term.map(|t| (sym, t)))
+            .collect::<Vec<_>>();
+        singletons.sort_by_key(|(_, term)| term.offset());
         singletons
             .into_iter()
-            .map(|term| diagnostic_from_singleton(term, self.kb))
+            .map(|(sym, term)| {
+                if let Value::Pattern(_) = term.value() {
+                    Diagnostic::Warning(
```

Comment:
Perhaps we should promote this to a full-blown error at this point?

Reply:
Was that the plan at some point? Off the top of my head I can't think of any reason why we shouldn't, but I'd rather make a follow-up task instead of stapling it onto this PR